### ISSUE TITLE
GitLab OpenID OAuth redirecte to a wrong URL

### ISSUE THREAD

**Issue by kestr31** [go-vikunja/desktop#8]
## Release specifications
- OS: Windows 11
- Verion: 0.20.3
- Session Type: Docker, Selfhosted, Reverse Proxy(NGINX)

## Issue
- I have set my Vikunja instance using docker on my server with OpenID OAuth.
- I am able to login to my server using Oauth set through the web frontend.
  - Settings I used is the [ame as the example google OAuth settings](https://vikunja.io/docs/openid-example-configurations/#google--google-workspace) except exact values used.
- However, I cannot login with the desktop application showing following error:

```error
An error has occurred<br/>
The redirect URI included is not valid.
```

## Guess of the cause
- At the first start, the api URL was shown as `localhost`.
- I think that mismatch in `AUTH_OPENID_REDIRECTURL` is causing a problem.

**Comment by kolaente** [comment]
Thanks for the kind words :slightly_smiling_face: 

The Vikunja Desktop app spins up its own local server to serve the frontend files, precisely to make redirects from an openid provider work. Because of that, you need to allow redirects back to `http://localhost` domains.

Does it work with the web frontend?

**Comment by verymilan** [comment]
I did whitelist the localhost url and then got stranded at `POST 400 /api/v1/auth/openid/tchncsauth/callback` 🤔

**Comment by kolaente** [comment]
Anything in the api logs?

**Comment by verymilan** [comment]
Hmm only `Response: {"error":"invalid_grant","error_description":"redirect_uri does not correspond"}` but the allowed redirect url is the same as with the frontend except `http://127.0.0.1:45735` (`http://localhost:45..` is allowed too, but the application makes a request containing `&redirect_uri=http://127.0.0.1:45735/auth/openid/...`) instead of the domain... unless the "developer mode" setting in the Zitadel server for allowing `http://` connections is broken, i don't know what could differ right now.

**Comment by kolaente** [comment]
Does Zitadel have an error message in their logs?

**Comment by verymilan** [comment]
Sadly even with debug loglevel there isn't anything to see about this problem. Its nginx shows a `POST /oauth/v2/token HTTP/2.0" 400`

**Comment by kolaente** [comment]
> the allowed redirect url is the same as with the frontend except http://127.0.0.1:45735

Do you have `http://127.0.0.1:45735` as an allowed redirect url? Or only the localhost one?

Does it work with the web frontend in a browser?

I don't know anything about Zitadel, maybe you need to provide the full url, so something like `[<frontend-url>](http://127.0.0.1:45735)/auth/openid/<auth key>` instead of the domain only.

**Comment by verymilan** [comment]
> Does it work with the web frontend in a browser?

Yes, it works fine in the browser. Right now allowed redirect urls are:
* https://todo.tchncs.de/auth/openid/tchncsauth
* http://127.0.0.1:45735/auth/openid/tchncsauth
* http://localhost:45735/auth/openid/tchncsauth

Just in case i have added http://ip/localhost:port as additional origin but it does not change...

**Comment by kolaente** [comment]
Huh that's strange. I've never seen this happen with other providers. Does Zitadel have an option to allow any redirect URL?

**Comment by verymilan** [comment]
Sadly not that i am aware of... but the redirect_uri is quite clear on the other hand: `redirect_uri=http://127.0.0.1:45735/auth/openid/tchncsauth&response_type[...]`
